plugins/pipelines/app/models/concerns/pipeline_plugin_concern.rb

```diff
@@ -0,0 +1,33 @@
+module PipelinePluginConcern
+  # Return true if any stages in the pipeline are marked production
+  def production?
+    super || Stage.find(next_stage_ids).any? { |s| s.production? }
+  end
+
+  def next_stage
+    return Stage.find(next_stage_ids.first) unless next_stage_ids.empty?
+    stages = project.stages.to_a
+    stages[stages.index(self) + 1]
+  end
+
+  # Ensure we don't have a circular pipeline:
+  #
+  # potential race-condition if 2 stages are saved at same time:
+  #   stageA saved with pipelines to stageB and stageC
+  #   stageC saved with pipeline to stageA   => will validate if stageA above hasn't been written to DB yet
+  def valid_pipeline?(origin_id = id)
+    Rails.logger.warn "valid_pipeline? #{origin_id} / #{id} - #{next_stage_ids}"
+    if next_stage_ids.any? { |next_id| next_id.to_i == origin_id.to_i }
+      errors[:base] << "Stage #{name} causes a circular pipeline with this stage"
+      return false
+    end
+
+    Stage.find(next_stage_ids).each do |stage|
```

Comment:
I think you can add all the errors and it will stop by itself -> less code + all bad stages show up